### ISSUE TITLE
`Relation::getRelated()` returns mixed — narrow to the related Model class

### ISSUE THREAD

**Issue by alies-dev** [psalm/laravel-psalm-plugin#760]
## Description

`Illuminate\Database\Eloquent\Relations\Relation::getRelated()` is declared to return the abstract `\Illuminate\Database\Eloquent\Model`, but at every concrete call site the related class is already known statically — either through the relation class itself (`HasMany<User, Post>` → `getRelated(): User`) or through the parent model's relation method (`$post->user()` → `BelongsTo<User, ...>` → `User`).

The plugin already understands Eloquent relation methods and emits typed relation subclasses from `belongsTo()`, `hasMany()`, `hasOne()`, `morphTo()`, etc. (see `src/Handlers/Eloquent/RelationMethodParser.php`, `ModelRelationshipPropertyHandler.php`). Extending that same knowledge to `->getRelated()` calls would close a small but consistently recurring cluster of false positives.

## Reproduction

Surfaced in a plugin v4.8.3 audit of [filamentphp/filament](https://github.com/filamentphp/filament) at `errorLevel=1`. Representative examples:

```php
// packages/forms/src/Components/ModalTableSelect.php:724
$record = $relationship->getRelated();          // MixedAssignment
                                                 // + MixedMethodCall on the next $record-> call

// packages/forms/src/Components/Repeater.php:1166
$relatedKeyName = $relationship->getRelated()->getKeyName();  // PossiblyNullReference + Mixed

// packages/support/src/Concerns/HasCellState.php:349
$record = $record->{$namePart}()->getRelated();  // MixedMethodCall

// packages/forms/src/Components/ModalTableSelect.php:266
->pluck(($relationship instanceof BelongsToMany)
    ? $relationship->getRelatedKeyName()
    : $relationship->getRelated()->getKeyName())
```

## Impact

From the v4.8.3 audit JSON, findings whose snippet contains `->getRelated()`:

| Count | Issue type | Shape |
| ---: | --- | --- |
| 11 | `MixedAssignment` | `$record = $relationship->getRelated();` |
| 11 | `MixedMethodCall` | `$record = $relationship->getRelated();` (same sites, cascade) |
| 6  | `MixedMethodCall` | `$record = $record->{$namePart}()->getRelated();` (dynamic relation — plugin gap is the same, just through a runtime name) |
| 3  | `MixedAssignment` | same dynamic-name form |
| 3  | `MixedMethodCall` | mixed argument to `pluck()` from `getRelated()->getKeyName()` |
| ~5 | mix of `MixedArgument`, `PossiblyNullReference`, `PropertyTypeCoercion` | chained calls on the result |

Roughly **40 findings**, of which ~22 are direct (the static-name form) and the remainder are cascaded. Real-world impact across Laravel packages is likely higher — any codebase that writes generic relation helpers (`$relation->getRelated()->newQuery()`, `$relation->getRelated()->getKeyName()`, etc.) hits this.

## Root cause

`ContainerResolver` / `ModelRegistrationHandler` resolve `belongsTo(User::class)` etc. to the specific relation generics, but no `MethodReturnTypeProviderInterface` is registered on `Relation` subclasses for `getRelated`. So `$relation->getRelated()` falls back to the declared `Model` (or `mixed` after Psalm's own narrowing loses the concrete type through the relation chain).

## Expected behavior

`$relation->getRelated()` should return the concrete related Model for the relation's second template parameter (or, depending on Laravel version, the first — see caveat below).

| Call site | Expected return |
| --- | --- |
| `\$post->user()->getRelated()` where `user(): BelongsTo<User, Post>` | `User` |
| `\$user->posts()->getRelated()` where `posts(): HasMany<Post, User>` | `Post` |
| `\$relation instanceof BelongsToMany && \$relation->getRelated()` on a `BelongsToMany<Role, User>` | `Role` |
| Generic `Relation<Model> \$r; \$r->getRelated();` (no narrowing possible) | `Model` (unchanged) |

## Proposed fix

Register a `MethodReturnTypeProviderInterface` on the `Relation` hierarchy. Rough shape:

```php
final class RelationGetRelatedHandler implements MethodReturnTypeProviderInterface
{
    public static function getClassLikeNames(): array
    {
        return [
            \Illuminate\Database\Eloquent\Relations\Relation::class,
            // Psalm walks subclasses automatically when the parent is listed,
            // so BelongsTo / HasMany / MorphTo / etc. are covered.
        ];
    }

    public static function getMethodReturnType(MethodReturnTypeProviderEvent \$event): ?Union
    {
        if (strtolower(\$event->getMethodNameLowercase()) !== 'getrelated') {
            return null;
        }

        \$calledOnType = \$event->getSource()->getNodeTypeProvider()->getType(\$event->getStmt()->var);
        if (\$calledOnType === null) {
            return null;
        }

        foreach (\$calledOnType->getAtomicTypes() as \$atomic) {
            if (!\$atomic instanceof TGenericObject) {
                continue;
            }
            // Relation<TRelatedModel, TParent> — first generic param is the related model.
            // Defensive: skip if the generic has fewer than one type param.
            \$params = \$atomic->type_params;
            if (\$params === []) {
                continue;
            }

            return \$params[0];
        }

        return null;
    }
}
```

Register it in `src/Plugin.php` alongside the existing Eloquent handlers.

## Caveats and edge cases

- **Template param order**. Laravel's `Relation<TRelatedModel, TDeclaringModel>` is the shape on current Laravel, but earlier versions and some subclass docblocks have the order reversed or added a third param (e.g. `BelongsToMany<TRelatedModel, TDeclaringModel, TPivot>`). The handler should read the first param for `HasOne/HasMany/BelongsTo/MorphTo/HasOneThrough/HasManyThrough/BelongsToMany/MorphMany/MorphOne` and ignore the rest. Add a regression test per relation type.
- **Dynamic relation names** (`\$record->{\$namePart}()->getRelated()`): the handler does not help here because the upstream relation type is already mixed. This is the already-acknowledged "dynamic relation" limitation tracked elsewhere; noted only so we do not over-claim the impact.
- **`\$relation->getParent()`**, **`\$relation->getQuery()`**, **`\$relation->getRelationExistenceQuery()`** and similar sibling methods have the same shape. It is worth covering them in the same handler while the logic is fresh. `getQuery()` returns `Builder<TRelated>`, `getParent()` returns `TDeclaring`.
- **`BelongsToMany::getRelatedKeyName()`**, **`getQualifiedRelatedKeyName()`**, etc. are getters returning strings — out of scope for this issue (no generic info needed, and the existing Laravel stubs already cover them).

## Where to look in the plugin

- `src/Handlers/Eloquent/RelationMethodParser.php` — produces the generic `Relation` return types from model methods. Confirms that the generic info is already available at the right place.
- `src/Handlers/Eloquent/ModelRelationshipPropertyHandler.php` — precedent for "given a relation generic, emit a Model type".
- `src/Plugin.php` — registration point for the new handler.

## Related

- #750 — `app(static::class)` resolution (fixed in v4.8.3)
- #751 — `filled()` / `blank()` stub (fixed in v4.8.3)
- #752 — `config()` FunctionReturnTypeProvider (pending)
- #758 — `::macro()` on Macroable classes (pending)
- #759 — Livewire `Testable::instance()` return typing (pending)

Same filamentphp/filament audit.
